### Домашнее задание 4 - 10 баллов

В этом задании вам предстоит дообучить трансформерную модель для NER-задачи в различных форматах:

1. Обучите NER-модель

- Загрузите набор данных [Collection5](https://github.com/natasha/corus?tab=readme-ov-file#load_ne5) - **1 балл**
- Разбейте набор данных на train/test части
- Дообучите модель [rubert-tiny2](https://huggingface.co/cointegrated/rubert-tiny2) на train-части корпуса для решения NER-задачи, сделайте замеры качества NER-метрик до и после дообучения - **2 балла**

2. Попробуйте улучшить качество модели следующими способами:
- Предварительно дообучите на train-части в MLM режиме, а потом дообучите на NER-задачу - **2 балла**
- Сгенерируйте синтетическую разметку* подходящего**, на ваш взгляд, новостного корпуса большой и умной моделью для русскоязычного NER***, а затем использовав ее для дообучения rubert-tiny2 вместе с основным набором данных - **2 балла**

3. Финально сравните результаты различных подходов - **1 балл**

*прогоните датасет через NER-модель, получите ее предсказания и используйте их в качестве резметки

**Можно использовать уже знакомый вам датасет lenta-ru, объем данных лучше взять от 10_000 текстов

***Например, можно взять модель модель DeepPavlov ner_collection3_bert. Инструкция по запуску есть в [документации](https://docs.deeppavlov.ai/en/master/features/models/NER.html)

**Общее**

- Принимаемые решения обоснованы (почему выбрана определенная архитектура/гиперпараметр/оптимизатор/преобразование и т.п.) - **1 балл**
- Обеспечена воспроизводимость решения: зафиксированы random_state, ноутбук воспроизводится от начала до конца без ошибок - **1 балл**

**Формат сдачи ДЗ**

- Каждая домашняя работа – PR в отдельную ветку **hw_n**, где **n** - номер домашней работы
- Добавить ментора и pacifikus в reviewers
- Дождаться ревью, если все ок – мержим в main
- Если не ок – вносим исправления и снова отправляем на ревью

# 1.Обучите NER-модель

## Загрузите набор данных Collection5

## Разбейте набор данных на train/test части

In [1]:
!pip install corus transformers datasets seqeval evaluate

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 4.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.7/83.7 kB 7.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.2/491.2 kB 36.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 8.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.9/183.9 kB 17.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 11.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 19.6 MB/s eta 0:00:00
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16162 sha256=c274a86bf6bba9f3d44f0c02bb76dde5c0291baed73238c05fe547ca8a6634b3
  Stored in directory: /root/.cache/pip/wheels/bc/92/f0/243288f899c2eacdfa8c5f9aede4c71a9bad0ee26a01dc5ead
Successfully built seqeval
  Attempting uninstall: fsspec
 

In [51]:
import random
import numpy as np
import torch
from corus import load_ne5
from corus import load_lenta
from sklearn.model_selection import train_test_split
from transformers import (
    AutoTokenizer,
    AutoModelForTokenClassification,
    AutoModelForMaskedLM,
    TrainingArguments,
    Trainer,
    DataCollatorForTokenClassification,
    DataCollatorForLanguageModeling,
    pipeline
)
from datasets import Dataset, DatasetDict, concatenate_datasets
import evaluate
import os
import random
import shutil
from tqdm import tqdm
import pandas as pd
import re

In [50]:
def set_seed(seed: int = 42) -> None:
    np.random.seed(seed)
    random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    # When running on the CuDNN backend, two further options must be set
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    # Set a fixed value for the hash seed
    os.environ["PYTHONHASHSEED"] = str(seed)
    print(f"Random seed set as {seed}")

set_seed(42)

Random seed set as 42


In [3]:
!wget http://www.labinform.ru/pub/named_entities/collection5.zip

--2025-04-13 06:46:41--  http://www.labinform.ru/pub/named_entities/collection5.zip
Resolving www.labinform.ru (www.labinform.ru)... 95.181.230.181
Connecting to www.labinform.ru (www.labinform.ru)|95.181.230.181|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1899530 (1.8M) [application/zip]
Saving to: ‘collection5.zip’

collection5.zip     100%[===================>]   1.81M  1.14MB/s    in 1.6s    

2025-04-13 06:46:44 (1.14 MB/s) - ‘collection5.zip’ saved [1899530/1899530]



In [4]:
%%capture
!unzip collection5.zip
!rm collection5.zip

# Скачиваем датасет
dir = 'Collection5/'
records = load_ne5(dir)
record = next(records)

In [5]:
record.text, record.spans[0]

('Президент России уволил 6 генералов МВД\r\n \r\nПрезидент России Дмитрий Медведев продолжает перестановки в руководстве МВД. Сегодня он уволил 6 генералов милиции.\r\n\r\nВ частности, указом президента отправлен в отставку начальник Центра обеспечения оперативно-служебной деятельности по противодействию экстремизму МВД генерал-майор милиции Владимир Булатов. Не взяли на работу в полицию и ряд региональных руководителей ведомства. Уволены начальник Управления на транспорте МВД по Северо-Кавказскому федеральному округу генерал-майор милиции Игорь Жуков, начальник Управления внутренних дел по Тверской области генерал-майор милиции Александр Куликов, начальник Управления внутренних дел по Тамбовской области генерал-майор милиции Владимир Фурсов.\r\n\r\nТакже президент отправил в отставку начальников 2 милицейских институтов: начальника Сибирского юридического института МВД РФ генерал-майора милиции Николая Михайлова и начальника Тюменского юридического института МВД генерал-майора милици

In [6]:
tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")

def load_annotations(txt_path):
    # Загружает текст и аннотации из BRAT файлов
    with open(txt_path, 'r', encoding='utf-8') as f:
        text = f.read()

    ann_path = txt_path.replace('.txt', '.ann')
    annotations = []

    if os.path.exists(ann_path):
        with open(ann_path, 'r', encoding='utf-8') as f:
            annotations = [
                {'label': parts[1].split()[0],
                 'start': int(parts[1].split()[1]),
                 'end': int(parts[1].split()[-1])}
                for line in f if line.startswith('T')
                for parts in [line.strip().split('\t')] if len(parts) >= 3
            ]

    return text, annotations

def align_tags(text, annotations):
    # Выравнивает теги с токенами
    tokens = tokenizer(text, return_offsets_mapping=True, add_special_tokens=False)
    labels = ['O'] * len(tokens['offset_mapping'])

    for ann in annotations:
        prefix = 'B-'
        for i, (start, end) in enumerate(tokens['offset_mapping']):
            if max(start, ann['start']) < min(end, ann['end']):
                labels[i] = f"{prefix}{ann['label']}"
                prefix = 'I-'

    return tokens.tokens(), labels

def process_directory(data_dir):
    # Обрабатывает все файлы в директории
    return [
        {
            'id': f[:-4],
            'text': text,
            'tokens': tokens,
            'ner_tags': tags
        }
        for f in os.listdir(data_dir) if f.endswith('.txt')
        for text, anns in [load_annotations(os.path.join(data_dir, f))]
        for tokens, tags in [align_tags(text, anns)]
    ]

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/401 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.08M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.74M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

In [7]:
data = process_directory("Collection5")
random.shuffle(data)

split = int(0.8 * len(data))
dataset = DatasetDict({
    'train': Dataset.from_list(data[:split]),
    'test': Dataset.from_list(data[split:])
})

print(f"Dataset ready: {len(data[:split])} train, {len(data[split:])} test examples")

Token indices sequence length is longer than the specified maximum sequence length for this model (2306 > 2048). Running this sequence through the model will result in indexing errors


Dataset ready: 800 train, 200 test examples


In [8]:
dataset

DatasetDict({
    train: Dataset({
        features: ['id', 'text', 'tokens', 'ner_tags'],
        num_rows: 800
    })
    test: Dataset({
        features: ['id', 'text', 'tokens', 'ner_tags'],
        num_rows: 200
    })
})

## Дообучите модель [rubert-tiny2](https://huggingface.co/cointegrated/rubert-tiny2) на train-части корпуса для решения NER-задачи, сделайте замеры качества NER-метрик до и после дообучения

Подготовка модели и токенизатора

In [42]:
# Загружаем токенизатор
tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")

# Получаем список уникальных тегов из train данных
label_list = sorted(set(tag for example in dataset['train'] for tag in example['ner_tags'] if tag != 'O'))
label_list = ['O'] + label_list  # Добавляем тег для не-сущностей

# Создаем маппинги
id2label = {i: label for i, label in enumerate(label_list)}
label2id = {label: i for i, label in enumerate(label_list)}

# Загружаем модель
model = AutoModelForTokenClassification.from_pretrained(
    "cointegrated/rubert-tiny2",
    num_labels=len(label_list),
    id2label=id2label,
    label2id=label2id
)

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Функция для вычисления метрик

In [31]:
def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    # Удаляем специальные токены (где label = -100)
    true_predictions = [
        [label_list[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [label_list[l] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]

    # Вычисляем метрики
    results = evaluate.load("seqeval").compute(
        predictions=true_predictions,
        references=true_labels
    )
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

Обучение модели

In [32]:
training_args = TrainingArguments(
    output_dir="rubert-tiny2-ner",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=10,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    metric_for_best_model="f1",
    seed=42,
    remove_unused_columns=False
)

# Функция для токенизации и выравнивания меток
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        is_split_into_words=True,
        truncation=True,
        padding="max_length",
        max_length=128
    )

    labels = []
    for i, label in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label2id[label[word_idx]])
            else:
                label_ids.append(-100)
            previous_word_idx = word_idx
        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

# Применяем токенизацию к датасету
tokenized_datasets = dataset.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=dataset["train"].column_names  # Удаляем оригинальные колонки
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["test"],
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
    data_collator=DataCollatorForTokenClassification(tokenizer=tokenizer)
)

trainer.train()
base_metrics = trainer.evaluate()

/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1611: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

<ipython-input-32-e6ccd01aed5d>:51: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Оценка модели ДО дообучения

In [49]:
model = AutoModelForTokenClassification.from_pretrained(
    "cointegrated/rubert-tiny2",
    num_labels=len(label_list),
    id2label=id2label,
    label2id=label2id
)

# Создаем временного Trainer для оценки модели до обучения
eval_trainer = Trainer(
    model=model,
    args=TrainingArguments(
        output_dir="./tmp_eval",
        per_device_eval_batch_size=16
    ),
    eval_dataset=tokenized_datasets["test"],
    compute_metrics=compute_metrics,
    data_collator=DataCollatorForTokenClassification(tokenizer=tokenizer)
)

print("Оценка модели до дообучения:")
initial_metrics = eval_trainer.evaluate()
print(f"F1: {initial_metrics['eval_f1']:.3f}")
print(f"Precision: {initial_metrics['eval_precision']:.3f}")
print(f"Recall: {initial_metrics['eval_recall']:.3f}")

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Оценка модели до дообучения:


F1: 0.010
Precision: 0.006
Recall: 0.040


# 2.Попробуйте улучшить качество модели следующими способами:

## Предварительно дообучите на train-части в MLM режиме, а потом дообучите на NER-задачу

Подготовка данных для MLM

In [33]:
# Создаем копию датасета, так как для MLM нам не нужны NER-теги
mlm_dataset = dataset.remove_columns("ner_tags")

# Инициализируем collator для MLM
mlm_data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=True,  # Включаем маскирование
    mlm_probability=0.15  # Вероятность маскирования токена
)

MLM-предобучение

In [34]:
# Загрузка модели и токенизатора
model_name = "cointegrated/rubert-tiny2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
mlm_model = AutoModelForMaskedLM.from_pretrained(model_name)

# Пподготовка данных для MLM
def prepare_mlm_data(examples):
    # Токенизируем текст (не разбитый на токены)
    return tokenizer(examples["text"], truncation=True, max_length=128)

# Применяем к датасету
mlm_dataset = dataset.map(
    prepare_mlm_data,
    batched=True,
    remove_columns=dataset["train"].column_names  # Удаляем исходные колонки
)

# Инициализация DataCollator для MLM
data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=True,
    mlm_probability=0.15
)

# Настройка обучения
training_args = TrainingArguments(
    output_dir="./mlm_pretrain",
    overwrite_output_dir=True,
    num_train_epochs=1,
    per_device_train_batch_size=16,
    save_steps=10_000,
    save_total_limit=2,
    prediction_loss_only=True,
    remove_unused_columns=False
)

# Создание Trainer
trainer = Trainer(
    model=mlm_model,
    args=training_args,
    train_dataset=mlm_dataset["train"],
    eval_dataset=None,
    data_collator=data_collator
)

# Запуск обучения
trainer.train()
print("MLM-предобучение завершено")

# Сохранение модели
mlm_model.save_pretrained("./mlm_pretrained")
tokenizer.save_pretrained("./mlm_pretrained")

Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

MLM-предобучение завершено


('./mlm_pretrained/tokenizer_config.json',
 './mlm_pretrained/special_tokens_map.json',
 './mlm_pretrained/vocab.txt',
 './mlm_pretrained/added_tokens.json',
 './mlm_pretrained/tokenizer.json')

Дообучение на NER

In [35]:
# Загрузка модели для NER с предобученными MLM-весами
ner_model = AutoModelForTokenClassification.from_pretrained(
    "./mlm_pretrained",
    num_labels=len(label_list),
    id2label=id2label,
    label2id=label2id
)

# Функция для токенизации NER-данных
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        is_split_into_words=True,
        truncation=True,
        padding="max_length",
        max_length=128
    )

    labels = []
    for i, label in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(label2id[label[word_idx]])
            else:
                label_ids.append(-100)
            previous_word_idx = word_idx
        labels.append(label_ids)

    tokenized_inputs["labels"] = labels
    return tokenized_inputs

# Подготовка данных
tokenized_datasets = dataset.map(
    tokenize_and_align_labels,
    batched=True,
    remove_columns=dataset["train"].column_names
)

# Настройки обучения для NER
ner_training_args = TrainingArguments(
    output_dir="ner_finetuned",
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=10,
    learning_rate=2e-5,
    weight_decay=0.01,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    metric_for_best_model="f1",
    seed=42,
    remove_unused_columns=False
)

# Инициализация Trainer
ner_trainer = Trainer(
    model=ner_model,
    args=ner_training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["test"],
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
    data_collator=DataCollatorForTokenClassification(tokenizer=tokenizer)
)

ner_trainer.train()
print("Дообучение на NER завершено")

# Оценка качества
mlm_ner_metrics = ner_trainer.evaluate()
print("\nРезультаты MLM + NER модели:")
print(f"F1-score: {mlm_ner_metrics['eval_f1']:.3f}")
print(f"Precision: {mlm_ner_metrics['eval_precision']:.3f}")
print(f"Recall: {mlm_ner_metrics['eval_recall']:.3f}")

Some weights of BertForTokenClassification were not initialized from the model checkpoint at ./mlm_pretrained and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1611: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
<ipython-input-35-bc8ba9e2e4df>:61: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  ner_trainer = Trainer(


/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Дообучение на NER завершено



Результаты MLM + NER модели:
F1-score: 0.297
Precision: 0.282
Recall: 0.314


## Сгенерируйте синтетическую разметку* подходящего**, на ваш взгляд, новостного корпуса большой и умной моделью для русскоязычного NER***, а затем использовав ее для дообучения rubert-tiny2 вместе с основным набором данных

Загрузка и подготовка данных Lenta.ru

In [16]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-04-13 06:49:53--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250413%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250413T064953Z&X-Amz-Expires=300&X-Amz-Signature=d78e8ac6edfd801795b9c0d3306db1e6ac878adc07f2ab9a6a041bb1b93f7ece&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-04-13 06:49:54--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-H

In [17]:
path = '/content/lenta-ru-news.csv.gz'
records = load_lenta(path)
df = pd.read_csv("lenta-ru-news.csv.gz").sample(10000, random_state=42)
next(records)

LentaRecord(
    url='https://lenta.ru/news/2018/12/14/cancer/',
    title='Названы регионы России с\xa0самой высокой смертностью от\xa0рака',
    text='Вице-премьер по социальным вопросам Татьяна Голикова рассказала, в каких регионах России зафиксирована наиболее высокая смертность от рака, сообщает РИА Новости. По словам Голиковой, чаще всего онкологические заболевания становились причиной смерти в Псковской, Тверской, Тульской и Орловской областях, а также в Севастополе. Вице-премьер напомнила, что главные факторы смертности в России — рак и болезни системы кровообращения. В начале года стало известно, что смертность от онкологических заболеваний среди россиян снизилась впервые за три года. По данным Росстата, в 2017 году от рака умерли 289 тысяч человек. Это на 3,5 процента меньше, чем годом ранее.',
    topic='Россия',
    tags='Общество',
    date=None
)

Генерация NER-разметки

In [19]:
# Инициализация NER-пайплайна
ner = pipeline(
    "ner",
    model="viktoroo/sberbank-rubert-base-collection3",
    device=0 if torch.cuda.is_available() else -1,
    aggregation_strategy="simple"
)

def add_ner_tags(text):
    # Токенизация текста
    words = text.split()[:200]  # Берем первые 200 слов для экономии памяти
    tags = ['O'] * len(words)

    try:
        # Обработка текста с ограничением длины
        entities = ner(text[:2000])

        for ent in entities:
            word = ent['word'].strip()
            if word in words:
                tags[words.index(word)] = f"B-{ent['entity_group']}"

    except Exception as e:
        print(f"Ошибка обработки текста: {str(e)}")

    return {'tokens': words, 'ner_tags': tags}

# Создаем датасет
samples = []
for text in tqdm(df['text'].head(10000), desc="Генерация NER"):
    samples.append(add_ner_tags(text))

dataset_synth = Dataset.from_list(samples)

Device set to use cuda:0
Генерация NER:  10%|▉         | 951/10000 [00:34<04:13, 35.67it/s]

Ошибка обработки текста: The size of tensor a (586) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  12%|█▏        | 1219/10000 [00:42<03:51, 37.90it/s]

Ошибка обработки текста: The size of tensor a (548) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  13%|█▎        | 1251/10000 [00:43<03:39, 39.94it/s]

Ошибка обработки текста: The size of tensor a (541) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  13%|█▎        | 1313/10000 [00:45<03:59, 36.26it/s]

Ошибка обработки текста: The size of tensor a (538) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  18%|█▊        | 1755/10000 [00:58<04:10, 32.86it/s]

Ошибка обработки текста: The size of tensor a (588) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  18%|█▊        | 1813/10000 [01:00<05:20, 25.56it/s]

Ошибка обработки текста: The size of tensor a (575) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  26%|██▌       | 2578/10000 [01:28<03:34, 34.57it/s]

Ошибка обработки текста: The size of tensor a (554) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  27%|██▋       | 2717/10000 [01:33<03:18, 36.68it/s]

Ошибка обработки текста: The size of tensor a (629) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  29%|██▉       | 2879/10000 [01:38<02:55, 40.56it/s]

Ошибка обработки текста: The size of tensor a (526) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  30%|███       | 3035/10000 [01:42<04:01, 28.88it/s]

Ошибка обработки текста: The size of tensor a (572) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  32%|███▏      | 3182/10000 [01:47<02:55, 38.89it/s]

Ошибка обработки текста: The size of tensor a (699) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  41%|████▏     | 4135/10000 [02:15<02:33, 38.29it/s]

Ошибка обработки текста: The size of tensor a (517) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  42%|████▏     | 4246/10000 [02:18<02:32, 37.63it/s]

Ошибка обработки текста: The size of tensor a (571) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  51%|█████     | 5052/10000 [02:42<02:02, 40.51it/s]

Ошибка обработки текста: The size of tensor a (568) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER:  93%|█████████▎| 9254/10000 [04:46<00:26, 28.56it/s]

Ошибка обработки текста: The size of tensor a (768) must match the size of tensor b (512) at non-singleton dimension 1


Генерация NER: 100%|██████████| 10000/10000 [05:08<00:00, 32.40it/s]


In [20]:
dataset_synth

Dataset({
    features: ['tokens', 'ner_tags'],
    num_rows: 10000
})

In [21]:
dataset_synth.save_to_disk("dataset_synth")
dataset_synth.to_csv("dataset_synth.csv")

Saving the dataset (0/1 shards):   0%|          | 0/10000 [00:00<?, ? examples/s]

Creating CSV from Arrow format:   0%|          | 0/10 [00:00<?, ?ba/s]

31523743

Подготовка данных

In [36]:
tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")

def tokenize_fn(examples):
    tokenized = tokenizer(examples["tokens"], is_split_into_words=True, truncation=True, max_length=256)
    labels = []
    for i, tags in enumerate(examples["ner_tags"]):
        word_ids = tokenized.word_ids(batch_index=i)
        labels.append([-100 if wid is None else label2id[tags[wid]] for wid in word_ids])
    tokenized["labels"] = labels
    return tokenized

dataset_train = dataset['train'].map(tokenize_fn, batched=True)
dataset_synth = dataset_synth.map(tokenize_fn, batched=True)
dataset_combined = concatenate_datasets([dataset_train, dataset_synth])

Map:   0%|          | 0/800 [00:00<?, ? examples/s]

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

Дообучение модели

In [38]:
model = AutoModelForTokenClassification.from_pretrained(
    "cointegrated/rubert-tiny2",
    num_labels=len(label_list),
    id2label=id2label,
    label2id=label2id
)

trainer = Trainer(
    model=model,
    args=TrainingArguments(
        output_dir="synth_ner_model",
        learning_rate=2e-5,
        per_device_train_batch_size=16,
        num_train_epochs=10,
        evaluation_strategy="epoch",
        save_strategy="epoch",
        load_best_model_at_end=True,
        metric_for_best_model="f1",
        seed=42
    ),
    train_dataset=dataset_combined,
    eval_dataset=dataset['test'].map(tokenize_fn, batched=True),
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
    data_collator=DataCollatorForTokenClassification(tokenizer)
)

print("До обучения:", trainer.evaluate())
trainer.train()
print("После обучения:", trainer.evaluate())

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1611: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Map:   0%|          | 0/200 [00:00<?, ? examples/s]

<ipython-input-38-86436ec2b3f6>:8: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


До обучения: {'eval_loss': 2.5059609413146973, 'eval_model_preparation_time': 0.0016, 'eval_precision': 0.005342246600388527, 'eval_recall': 0.04192825112107623, 'eval_f1': 0.00947699168862761, 'eval_accuracy': 0.045554416288267174, 'eval_runtime': 4.7889, 'eval_samples_per_second': 41.764, 'eval_steps_per_second': 5.22}


/usr/local/lib/python3.11/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


После обучения: {'eval_loss': 0.470948189496994, 'eval_model_preparation_time': 0.0016, 'eval_precision': 0.2690331620737973, 'eval_recall': 0.25829596412556055, 'eval_f1': 0.26355525051475637, 'eval_accuracy': 0.843687877963548, 'eval_runtime': 2.4813, 'eval_samples_per_second': 80.604, 'eval_steps_per_second': 10.076, 'epoch': 10.0}


In [39]:
synth_metrics = trainer.evaluate()
print("\nРезультаты модели с синтетическими данными:")
print(f"F1-score: {synth_metrics['eval_f1']:.3f}")
print(f"Precision: {synth_metrics['eval_precision']:.3f}")
print(f"Recall: {synth_metrics['eval_recall']:.3f}")


Результаты модели с синтетическими данными:
F1-score: 0.264
Precision: 0.269
Recall: 0.258


# 3.Финально сравните результаты различных подходов

In [41]:
# Финальное сравнение моделей
print("\n" + "="*50)
print("ФИНАЛЬНОЕ СРАВНЕНИЕ РЕЗУЛЬТАТОВ")
print("="*50)

# 1. Базовая модель
print("\n1. Базовая модель:")
print(f"F1-score: {base_metrics['eval_f1']:.3f}")
print(f"Precision: {base_metrics['eval_precision']:.3f}")
print(f"Recall: {base_metrics['eval_recall']:.3f}")
print(f"Accuracy: {base_metrics['eval_accuracy']:.3f}")

# 2. MLM + NER
print("\n2. MLM + NER:")
print(f"F1-score: {mlm_ner_metrics['eval_f1']:.3f} (Δ{mlm_ner_metrics['eval_f1']-base_metrics['eval_f1']:+.3f})")
print(f"Precision: {mlm_ner_metrics['eval_precision']:.3f} (Δ{mlm_ner_metrics['eval_precision']-base_metrics['eval_precision']:+.3f})")
print(f"Recall: {mlm_ner_metrics['eval_recall']:.3f} (Δ{mlm_ner_metrics['eval_recall']-base_metrics['eval_recall']:+.3f})")
print(f"Accuracy: {mlm_ner_metrics['eval_accuracy']:.3f} (Δ{mlm_ner_metrics['eval_accuracy']-base_metrics['eval_accuracy']:+.3f})")

# 3. Синтетика + NER
print("\n3. Синтетика + NER:")
print(f"F1-score: {synth_metrics['eval_f1']:.3f} (Δ{synth_metrics['eval_f1']-base_metrics['eval_f1']:+.3f})")
print(f"Precision: {synth_metrics['eval_precision']:.3f} (Δ{synth_metrics['eval_precision']-base_metrics['eval_precision']:+.3f})")
print(f"Recall: {synth_metrics['eval_recall']:.3f} (Δ{synth_metrics['eval_recall']-base_metrics['eval_recall']:+.3f})")
print(f"Accuracy: {synth_metrics['eval_accuracy']:.3f} (Δ{synth_metrics['eval_accuracy']-base_metrics['eval_accuracy']:+.3f})")

# Визуализация сравнения
print("\n" + "="*50)
print("СВОДНАЯ ТАБЛИЦА РЕЗУЛЬТАТОВ")
print("="*50)
print(f"{'Модель':<20} {'F1':<8} {'Precision':<10} {'Recall':<10} {'Accuracy':<10}")
print("-"*50)
print(f"{'Базовая':<20} {base_metrics['eval_f1']:.3f} {base_metrics['eval_precision']:.3f} {base_metrics['eval_recall']:.3f} {base_metrics['eval_accuracy']:.3f}")
print(f"{'MLM + NER':<20} {mlm_ner_metrics['eval_f1']:.3f} {mlm_ner_metrics['eval_precision']:.3f} {mlm_ner_metrics['eval_recall']:.3f} {mlm_ner_metrics['eval_accuracy']:.3f}")
print(f"{'Синтетика + NER':<20} {synth_metrics['eval_f1']:.3f} {synth_metrics['eval_precision']:.3f} {synth_metrics['eval_recall']:.3f} {synth_metrics['eval_accuracy']:.3f}")


ФИНАЛЬНОЕ СРАВНЕНИЕ РЕЗУЛЬТАТОВ

1. Базовая модель:
F1-score: 0.284
Precision: 0.266
Recall: 0.305
Accuracy: 0.826

2. MLM + NER:
F1-score: 0.297 (Δ+0.013)
Precision: 0.282 (Δ+0.016)
Recall: 0.314 (Δ+0.009)
Accuracy: 0.829 (Δ+0.003)

3. Синтетика + NER:
F1-score: 0.264 (Δ-0.020)
Precision: 0.269 (Δ+0.003)
Recall: 0.258 (Δ-0.046)
Accuracy: 0.844 (Δ+0.018)

СВОДНАЯ ТАБЛИЦА РЕЗУЛЬТАТОВ
Модель               F1       Precision  Recall     Accuracy  
--------------------------------------------------
Базовая              0.284 0.266 0.305 0.826
MLM + NER            0.297 0.282 0.314 0.829
Синтетика + NER      0.264 0.269 0.258 0.844


**Выводы:**

1. **Сравнение моделей**:
   - **MLM + NER** показала небольшое улучшение по всем метрикам (+0.013 F1, +0.016 Precision, +0.009 Recall) по сравнению с базовой моделью. Это говорит о том, что предобучение в MLM-режиме помогает модели лучше адаптироваться к данным, но эффект незначительный.
   - **Синтетика + NER** дала противоречивые результаты. Улучшились Precision (+0.003) и Accuracy (+0.018). Однако F1 и Recall ухудшились (-0.020 и -0.046), что означает, что модель стала пропускать часть сущностей и хуже балансировать между Precision и Recall.  

2. **Низкие значения метрик**:
   - Все модели показали невысокие результаты (F1 ~0.26–0.30), что связано с малым количеством эпох обучения (10). Так как NER-модели требуют больше времени для сходимости, особенно при дообучении на небольших датасетах.  

3. **Интерпретация результатов**:
   - **MLM + NER** — наиболее стабильный подход, но требует доработки (например больше эпох MLM-предобучения).  
   - **Синтетика + NER** не оправдала ожиданий, вероятно из-за шума в автоматической разметке. Для улучшения можно попробовать:  
     - Ручную постобработку синтетических меток.  
     - Использовать более качественную модель для генерации разметки (например DeepPavlov).  

**Общий итог**:
   - Лучший результат у **MLM + NER**, но все модели требуют дальнейшего улучшения.  
   - Для финального решения стоит сосредоточиться на увеличении эпох и проверке качества разметки.

